File: src/css/properties/display.zig
const std = @import("std");
const bun = @import("root").bun;
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayListUnmanaged;

pub const css = @import("../css_parser.zig");

const SmallList = css.SmallList;
const Printer = css.Printer;
const PrintErr = css.PrintErr;
const Error = css.Error;

const ContainerName = css.css_rules.container.ContainerName;

const LengthPercentage = css.css_values.length.LengthPercentage;
const CustomIdent = css.css_values.ident.CustomIdent;
const CSSString = css.css_values.string.CSSString;
const CSSNumber = css.css_values.number.CSSNumber;
const LengthPercentageOrAuto = css.css_values.length.LengthPercentageOrAuto;
const Size2D = css.css_values.size.Size2D;
const DashedIdent = css.css_values.ident.DashedIdent;
const Image = css.css_values.image.Image;
const CssColor = css.css_values.color.CssColor;
const Ratio = css.css_values.ratio.Ratio;
const Length = css.css_values.length.LengthValue;
const Rect = css.css_values.rect.Rect;
const NumberOrPercentage = css.css_values.percentage.NumberOrPercentage;
const CustomIdentList = css.css_values.ident.CustomIdentList;

/// A value for the [display](https://drafts.csswg.org/css-display-3/#the-display-properties) property.
pub const Display = union(enum) {
    /// A display keyword.
    keyword: DisplayKeyword,
    /// The inside and outside display values.
    pair: DisplayPair,

    pub usingnamespace css.DeriveParse(@This());
    pub usingnamespace css.DeriveToCss(@This());

    pub fn deepClone(this: *const @This(), allocator: std.mem.Allocator) @This() {
        return css.implementDeepClone(@This(), this, allocator);
    }

    pub fn hash(this: *const @This(), hasher: *std.hash.Wyhash) void {
        return css.implementHash(@This(), this, hasher);
    }

    pub fn eql(lhs: *const @This(), rhs: *const @This()) bool {
        return css.implementEql(@This(), lhs, rhs);
    }
};

/// A value for the [visibility](https://drafts.csswg.org/css-display-3/#visibility) property.
pub const Visibility = enum {
    /// The element is visible.
    visible,
    /// The element is hidden.
    hidden,
    /// The element is collapsed.
    collapse,

    pub usingnamespace css.DefineEnumProperty(@This());
};

/// A `display` keyword.
///
/// See [Display](Display).
pub const DisplayKeyword = enum {
    none,
    contents,
    @"table-row-group",
    @"table-header-group",
    @"table-footer-group",
    @"table-row",
    @"table-cell",
    @"table-column-group",
    @"table-column",
    @"table-caption",
    @"ruby-base",
    @"ruby-text",
    @"ruby-base-container",
    @"ruby-text-container",

    pub usingnamespace css.DefineEnumProperty(@This());
};

/// A pair of inside and outside display values, as used in the `display` property.
///
/// See [Display](Display).
pub const DisplayPair = struct {
    /// The outside display value.
    outside: DisplayOutside,
    /// The inside display value.
    inside: DisplayInside,
    /// Whether this is a list item.
    is_list_item: bool,

    pub fn parse(input: *css.Parser) css.Result(@This()) {
        var list_item = false;
        var outside: ?DisplayOutside = null;
        var inside: ?DisplayInside = null;

        while (true) {
            if (input.tryParse(css.Parser.expectIdentMatching, .{"list-item"}).isOk()) {
                list_item = true;
                continue;
            }

            if (outside == null) {
                if (input.tryParse(DisplayOutside.parse, .{}).asValue()) |o| {
                    outside = o;
                    continue;
                }
            }

            if (inside == null) {
                if (input.tryParse(DisplayInside.parse, .{}).asValue()) |i| {
                    inside = i;
                    continue;
                }
            }

            break;
        }

        if (list_item or inside != null or outside != null) {
            const final_inside: DisplayInside = inside orelse DisplayInside.flow;
            const final_outside: DisplayOutside = outside orelse switch (final_inside) {
                // "If <display-outside> is omitted, the element’s outside display type
                // defaults to block — except for ruby, which defaults to inline."
                // https://drafts.csswg.org/css-display/#inside-model
                .ruby => .@"inline",
                else => .block,
            };

            if (list_item and !(final_inside == .flow or final_inside == .flow_root)) {
                return .{ .err = input.newCustomError(.invalid_declaration) };
            }

            return .{ .result = .{
                .outside = final_outside,
                .inside = final_inside,
                .is_list_item = list_item,
            } };
        }

        const location = input.currentSourceLocation();
        const ident = switch (input.expectIdent()) {
            .result => |v| v,
            .err => |e| return .{ .err = e },
        };

        const displayIdentMap = bun.ComptimeStringMap(DisplayPair, .{
            .{ "inline-block", DisplayPair{ .outside = .@"inline", .inside = .flow_root, .is_list_item = false } },
            .{ "inline-table", DisplayPair{ .outside = .@"inline", .inside = .table, .is_list_item = false } },
            .{ "inline-flex", DisplayPair{ .outside = .@"inline", .inside = .{ .flex = css.VendorPrefix{ .none = true } }, .is_list_item = false } },
            .{ "-webkit-inline-flex", DisplayPair{ .outside = .@"inline", .inside = .{ .flex = css.VendorPrefix{ .webkit = true } }, .is_list_item = false } },
            .{ "-ms-inline-flexbox", DisplayPair{ .outside = .@"inline", .inside = .{ .flex = css.VendorPrefix{ .ms = true } }, .is_list_item = false } },
            .{ "-webkit-inline-box", DisplayPair{ .outside = .@"inline", .inside = .{ .box = css.VendorPrefix{ .webkit = true } }, .is_list_item = false } },
            .{ "-moz-inline-box", DisplayPair{ .outside = .@"inline", .inside = .{ .box = css.VendorPrefix{ .moz = true } }, .is_list_item = false } },
            .{ "inline-grid", DisplayPair{ .outside = .@"inline", .inside = .grid, .is_list_item = false } },
        });
        if (displayIdentMap.getASCIIICaseInsensitive(ident)) |pair| {
            return .{ .result = pair };
        }

        return .{ .err = location.newUnexpectedTokenError(.{ .ident = ident }) };
    }

    pub fn toCss(this: *const DisplayPair, comptime W: type, dest: *css.Printer(W)) css.PrintErr!void {
        if (this.outside == .@"inline" and this.inside == .flow_root and !this.is_list_item) {
            return dest.writeStr("inline-block");
        } else if (this.outside == .@"inline" and this.inside == .table and !this.is_list_item) {
            return dest.writeStr("inline-table");
        } else if (this.outside == .@"inline" and this.inside == .flex and !this.is_list_item) {
            try this.inside.flex.toCss(W, dest);
            if (this.inside.flex.eql(css.VendorPrefix{ .ms = true })) {
                return dest.writeStr("inline-flexbox");
            } else {
                return dest.writeStr("inline-flex");
            }
        } else if (this.outside == .@"inline" and this.inside == .box and !this.is_list_item) {
            try this.inside.box.toCss(W, dest);
            return dest.writeStr("inline-box");
        } else if (this.outside == .@"inline" and this.inside == .grid and !this.is_list_item) {
            return dest.writeStr("inline-grid");
        } else {
            const default_outside: DisplayOutside = switch (this.inside) {
                .ruby => .@"inline",
                else => .block,
            };

            var needs_space = false;
            if (!this.outside.eql(&default_outside) or (this.inside.eql(&DisplayInside{ .flow = {} }) and !this.is_list_item)) {
                try this.outside.toCss(W, dest);
                needs_space = true;
            }

            if (!this.inside.eql(&DisplayInside{ .flow = {} })) {
                if (needs_space) {
                    try dest.writeChar(' ');
                }
                try this.inside.toCss(W, dest);
                needs_space = true;
            }

            if (this.is_list_item) {
                if (needs_space) {
                    try dest.writeChar(' ');
                }
                try dest.writeStr("list-item");
            }
        }
    }

    pub fn eql(lhs: *const @This(), rhs: *const @This()) bool {
        return css.implementEql(@This(), lhs, rhs);
    }
};

/// A [`<display-outside>`](https://drafts.csswg.org/css-display-3/#typedef-display-outside) value.
pub const DisplayOutside = enum {
    block,
    @"inline",
    @"run-in",

    pub usingnamespace css.DefineEnumProperty(@This());
};

/// A [`<display-inside>`](https://drafts.csswg.org/css-display-3/#typedef-display-inside) value.
pub const DisplayInside = union(enum) {
    flow,
    flow_root,
    table,
    flex: css.VendorPrefix,
    box: css.VendorPrefix,
    grid,
    ruby,

    pub fn parse(input: *css.Parser) css.Result(@This()) {
        const displayInsideMap = bun.ComptimeStringMap(DisplayInside, .{
            .{ "flow", DisplayInside.flow },
            .{ "flow-root", DisplayInside.flow_root },
            .{ "table", .table },
            .{ "flex", .{ .flex = css.VendorPrefix{ .none = true } } },
            .{ "-webkit-flex", .{ .flex = css.VendorPrefix{ .webkit = true } } },
            .{ "-ms-flexbox", .{ .flex = css.VendorPrefix{ .ms = true } } },
            .{ "-webkit-box", .{ .box = css.VendorPrefix{ .webkit = true } } },
            .{ "-moz-box", .{ .box = css.VendorPrefix{ .moz = true } } },
            .{ "grid", .grid },
            .{ "ruby", .ruby },
        });

        const location = input.currentSourceLocation();
        const ident = switch (input.expectIdent()) {
            .result => |v| v,
            .err => |e| return .{ .err = e },
        };

        if (displayInsideMap.getASCIIICaseInsensitive(ident)) |value| {
            return .{ .result = value };
        }

        return .{ .err = location.newUnexpectedTokenError(.{ .ident = ident }) };
    }

    pub fn toCss(this: *const DisplayInside, comptime W: type, dest: *css.Printer(W)) css.PrintErr!void {
        switch (this.*) {
            .flow => try dest.writeStr("flow"),
            .flow_root => try dest.writeStr("flow-root"),
            .table => try dest.writeStr("table"),
            .flex => |prefix| {
                try prefix.toCss(W, dest);
                if (prefix.eql(css.VendorPrefix{ .ms = true })) {
                    try dest.writeStr("flexbox");
                } else {
                    try dest.writeStr("flex");
                }
            },
            .box => |prefix| {
                try prefix.toCss(W, dest);
                try dest.writeStr("box");
            },
            .grid => try dest.writeStr("grid"),
            .ruby => try dest.writeStr("ruby"),
        }
    }

    pub fn eql(lhs: *const @This(), rhs: *const @This()) bool {
        return css.implementEql(@This(), lhs, rhs);
    }
};

File: src/css/properties/display.rs
#![allow(unused_imports, dead_code)]
#![warn(unused_must_use)]
use crate as css;
use crate::{Parser, PrintErr, Printer, VendorPrefix};

/// A value for the [display](https://drafts.csswg.org/css-display-3/#the-display-properties) property.
#[derive(Clone, PartialEq, Eq, Hash)]
pub enum Display {
    /// A display keyword.
    Keyword(DisplayKeyword),
    /// The inside and outside display values.
    Pair(DisplayPair),
}

// PORT NOTE: Zig `DeriveParse`/`DeriveToCss` for a 2-payload union(enum) tries each
// payload's `parse` in declaration order; `toCss` dispatches to the active payload.
impl Display {
    pub fn parse(input: &mut Parser) -> css::Result<Self> {
        if let Ok(kw) = input.try_parse(DisplayKeyword::parse) {
            return Ok(Display::Keyword(kw));
        }
        DisplayPair::parse(input).map(Display::Pair)
    }

    pub fn to_css(&self, dest: &mut Printer) -> Result<(), PrintErr> {
        match self {
            Display::Keyword(kw) => kw.to_css(dest),
            Display::Pair(p) => p.to_css(dest),
        }
    }

    pub fn deep_clone(&self, _bump: &bun_alloc::Arena) -> Self {
        // All payloads are Copy.
        self.clone()
    }
}

/// A value for the [visibility](https://drafts.csswg.org/css-display-3/#visibility) property.
#[derive(Clone, Copy, PartialEq, Eq, Hash, css::DefineEnumProperty)]
pub enum Visibility {
    /// The element is visible.
    Visible,
    /// The element is hidden.
    Hidden,
    /// The element is collapsed.
    Collapse,
}

/// A `display` keyword.
///
/// See [Display](Display).
#[derive(Clone, Copy, PartialEq, Eq, Hash, css::DefineEnumProperty)]
pub enum DisplayKeyword {
    None,
    Contents,
    TableRowGroup,
    TableHeaderGroup,
    TableFooterGroup,
    TableRow,
    TableCell,
    TableColumnGroup,
    TableColumn,
    TableCaption,
    RubyBase,
    RubyText,
    RubyBaseContainer,
    RubyTextContainer,
}

/// A pair of inside and outside display values, as used in the `display` property.
///
/// See [Display](Display).
#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub struct DisplayPair {
    /// The outside display value.
    pub outside: DisplayOutside,
    /// The inside display value.
    pub inside: DisplayInside,
    /// Whether this is a list item.
    pub is_list_item: bool,
}

impl DisplayPair {
    pub fn parse(input: &mut Parser) -> css::Result<Self> {
        let mut list_item = false;
        let mut outside: Option<DisplayOutside> = None;
        let mut inside: Option<DisplayInside> = None;

        loop {
            if input
                .try_parse(|i| i.expect_ident_matching(b"list-item"))
                .is_ok()
            {
                list_item = true;
                continue;
            }

            if outside.is_none() {
                if let Ok(o) = input.try_parse(DisplayOutside::parse) {
                    outside = Some(o);
                    continue;
                }
            }

            if inside.is_none() {
                if let Ok(i) = input.try_parse(DisplayInside::parse) {
                    inside = Some(i);
                    continue;
                }
            }

            break;
        }

        if list_item || inside.is_some() || outside.is_some() {
            let final_inside: DisplayInside = inside.unwrap_or(DisplayInside::Flow);
            let final_outside: DisplayOutside = outside.unwrap_or(match final_inside {
                // "If <display-outside> is omitted, the element’s outside display type
                // defaults to block — except for ruby, which defaults to inline."
                // https://drafts.csswg.org/css-display/#inside-model
                DisplayInside::Ruby => DisplayOutside::Inline,
                _ => DisplayOutside::Block,
            });

            if list_item && !matches!(final_inside, DisplayInside::Flow | DisplayInside::FlowRoot) {
                return Err(input.new_custom_error(css::ParserError::invalid_declaration));
            }

            return Ok(DisplayPair {
                outside: final_outside,
                inside: final_inside,
                is_list_item: list_item,
            });
        }

        let location = input.current_source_location();
        let ident = input.expect_ident_cloned()?;

        // PORT NOTE: Zig used `bun.ComptimeStringMap(..).getASCIIICaseInsensitive`.
        // 8 keys → if-chain over `eql_case_insensitive_ascii::<true>` (phf values
        // would have to be const-eval, and `VendorPrefix` bitflags are not).
        use bun_core::eql_case_insensitive_ascii as eq;
        let inside = if eq(ident, b"inline-block", true) {
            DisplayInside::FlowRoot
        } else if eq(ident, b"inline-table", true) {
            DisplayInside::Table
        } else if eq(ident, b"inline-flex", true) {
            DisplayInside::Flex(VendorPrefix::NONE)
        } else if eq(ident, b"-webkit-inline-flex", true) {
            DisplayInside::Flex(VendorPrefix::WEBKIT)
        } else if eq(ident, b"-ms-inline-flexbox", true) {
            DisplayInside::Flex(VendorPrefix::MS)
        } else if eq(ident, b"-webkit-inline-box", true) {
            DisplayInside::Box(VendorPrefix::WEBKIT)
        } else if eq(ident, b"-moz-inline-box", true) {
            DisplayInside::Box(VendorPrefix::MOZ)
        } else if eq(ident, b"inline-grid", true) {
            DisplayInside::Grid
        } else {
            return Err(location.new_unexpected_token_error(css::Token::Ident(ident)));
        };
        Ok(DisplayPair {
            outside: DisplayOutside::Inline,
            inside,
            is_list_item: false,
        })
    }

    pub fn to_css(&self, dest: &mut Printer) -> Result<(), PrintErr> {
        // PORT NOTE: reshaped Zig if-else chain into match for tagged-union payload extraction.
        match (self.outside, &self.inside, self.is_list_item) {
            (DisplayOutside::Inline, DisplayInside::FlowRoot, false) => {
                return dest.write_str("inline-block");
            }
            (DisplayOutside::Inline, DisplayInside::Table, false) => {
                return dest.write_str("inline-table");
            }
            (DisplayOutside::Inline, DisplayInside::Flex(prefix), false) => {
                prefix.to_css(dest)?;
                if *prefix == VendorPrefix::MS {
                    return dest.write_str("inline-flexbox");
                } else {
                    return dest.write_str("inline-flex");
                }
            }
            (DisplayOutside::Inline, DisplayInside::Box(prefix), false) => {
                prefix.to_css(dest)?;
                return dest.write_str("inline-box");
            }
            (DisplayOutside::Inline, DisplayInside::Grid, false) => {
                return dest.write_str("inline-grid");
            }
            _ => {
                let default_outside: DisplayOutside = match self.inside {
                    DisplayInside::Ruby => DisplayOutside::Inline,
                    _ => DisplayOutside::Block,
                };

                let mut needs_space = false;
                if self.outside != default_outside
                    || (self.inside == DisplayInside::Flow && !self.is_list_item)
                {
                    self.outside.to_css(dest)?;
                    needs_space = true;
                }

                if self.inside != DisplayInside::Flow {
                    if needs_space {
                        dest.write_char(b' ')?;
                    }
                    self.inside.to_css(dest)?;
                    needs_space = true;
                }

                if self.is_list_item {
                    if needs_space {
                        dest.write_char(b' ')?;
                    }
                    dest.write_str("list-item")?;
                }
                Ok(())
            }
        }
    }
}

/// A [`<display-outside>`](https://drafts.csswg.org/css-display-3/#typedef-display-outside) value.
#[derive(Clone, Copy, PartialEq, Eq, Hash, css::DefineEnumProperty)]
pub enum DisplayOutside {
    Block,
    Inline,
    RunIn,
}

/// A [`<display-inside>`](https://drafts.csswg.org/css-display-3/#typedef-display-inside) value.
#[derive(Clone, Copy, PartialEq, Eq, Hash)]
pub enum DisplayInside {
    Flow,
    FlowRoot,
    Table,
    Flex(VendorPrefix),
    Box(VendorPrefix),
    Grid,
    Ruby,
}

impl DisplayInside {
    pub fn parse(input: &mut Parser) -> css::Result<Self> {
        let location = input.current_source_location();
        let ident = input.expect_ident_cloned()?;

        // PORT NOTE: Zig used `bun.ComptimeStringMap(..).getASCIIICaseInsensitive`.
        // 10 keys → if-chain over `eql_case_insensitive_ascii::<true>`.
        use bun_core::eql_case_insensitive_ascii as eq;
        Ok(if eq(ident, b"flow", true) {
            DisplayInside::Flow
        } else if eq(ident, b"flow-root", true) {
            DisplayInside::FlowRoot
        } else if eq(ident, b"table", true) {
            DisplayInside::Table
        } else if eq(ident, b"flex", true) {
            DisplayInside::Flex(VendorPrefix::NONE)
        } else if eq(ident, b"-webkit-flex", true) {
            DisplayInside::Flex(VendorPrefix::WEBKIT)
        } else if eq(ident, b"-ms-flexbox", true) {
            DisplayInside::Flex(VendorPrefix::MS)
        } else if eq(ident, b"-webkit-box", true) {
            DisplayInside::Box(VendorPrefix::WEBKIT)
        } else if eq(ident, b"-moz-box", true) {
            DisplayInside::Box(VendorPrefix::MOZ)
        } else if eq(ident, b"grid", true) {
            DisplayInside::Grid
        } else if eq(ident, b"ruby", true) {
            DisplayInside::Ruby
        } else {
            return Err(location.new_unexpected_token_error(css::Token::Ident(ident)));
        })
    }

    pub fn to_css(&self, dest: &mut Printer) -> Result<(), PrintErr> {
        match self {
            DisplayInside::Flow => dest.write_str("flow"),
            DisplayInside::FlowRoot => dest.write_str("flow-root"),
            DisplayInside::Table => dest.write_str("table"),
            DisplayInside::Flex(prefix) => {
                prefix.to_css(dest)?;
                if *prefix == VendorPrefix::MS {
                    dest.write_str("flexbox")
                } else {
                    dest.write_str("flex")
                }
            }
            DisplayInside::Box(prefix) => {
                prefix.to_css(dest)?;
                dest.write_str("box")
            }
            DisplayInside::Grid => dest.write_str("grid"),
            DisplayInside::Ruby => dest.write_str("ruby"),
        }
    }
}

// ported from: src/css/properties/display.zig
